File: lib/games/mafia/mafia_server.dart
import 'dart:async';
import 'dart:convert';

import '../../social/host_server.dart';
import '../../tutorials/tutorial_content.dart';
import 'mafia_browser.dart';
import 'mafia_engine.dart';
import 'mafia_role.dart';

/// Wraps [HostServer] with Mafia-specific routing. Owns the engine, fans
/// out the right private/public messages, and converts incoming
/// guest commands into engine calls.
class MafiaServer {
  MafiaServer({HostServer? server, this.hostName = 'Host'})
      : _server = server ?? HostServer(html: mafiaBrowserHtml);

  final HostServer _server;
  final String hostName;
  final MafiaEngine engine = MafiaEngine();

  /// Flutter host plays as this player; not connected over WebSocket.
  static const String hostId = 'host';
  GuestId? _hostingFromGuest;
  final Map<GuestId, String> _guestToPlayer = {};
  final Map<String, GuestId> _playerToGuest = {};

  final _stateChanges = StreamController<void>.broadcast();
  Stream<void> get onStateChange => _stateChanges.stream;

  StreamSubscription<GuestMessage>? _msgSub;
  StreamSubscription<GuestId>? _leaveSub;

  Future<Uri?> start() async {
    engine.addPlayer(id: hostId, name: hostName, isHost: true);
    final uri = await _server.start();
    _msgSub = _server.onGuestMessage.listen(_onMessage);
    _leaveSub = _server.onGuestLeave.listen(_onLeave);
    _emit();
    return uri;
  }

  Future<void> stop() async {
    await _msgSub?.cancel();
    await _leaveSub?.cancel();
    await _server.stop();
    await _stateChanges.close();
  }

  /// Number of currently connected browser/app guests.
  int get guestCount => _server.guestCount;

  // ---- Host-side actions (called from the Flutter host UI) -------------

  void hostStart() {
    if (!engine.canStart) return;
    engine.start();
    _broadcastPhase();
    _sendRolesPrivately();
    _emit();
  }

  void hostNightAction(String targetId) =>
      _applyNightAction(hostId, targetId);
  void hostDayVote(String? targetId) => _applyDayVote(hostId, targetId);

  void hostCallTutorialVote() => _openTutorialVote();
  void hostTutorialVote(bool yes) => _submitTutorialVote(hostId, yes);
  void hostDismissTutorial() {
    engine.tutorialVote.markShown();
    _broadcastTutorialState();
    _emit();
  }

  void advanceFromReveal() {
    engine.advanceToDayVote();
    if (engine.phase == MafiaPhase.gameOver) {
      _broadcastGameOver();
    } else {
      _broadcastPhase();
    }
    _emit();
  }

  // ---- Guest message handling ------------------------------------------

  void _onMessage(GuestMessage msg) {
    final json = msg.asJson;
    final type = json['type'] as String?;
    switch (type) {
      case 'join':
        _handleJoin(msg.from, json);
      case 'night_action':
        final pid = _guestToPlayer[msg.from];
        if (pid != null) _applyNightAction(pid, json['targetId'] as String);
      case 'vote':
        final pid = _guestToPlayer[msg.from];
        if (pid != null) _applyDayVote(pid, json['targetId'] as String?);
      case 'call_tutorial_vote':
        _openTutorialVote();
      case 'tutorial_vote':
        final pid = _guestToPlayer[msg.from];
        if (pid != null) _submitTutorialVote(pid, json['yes'] as bool);
    }
  }

  void _onLeave(GuestId guest) {
    final pid = _guestToPlayer.remove(guest);
    if (pid == null) return;
    _playerToGuest.remove(pid);
    engine.removePlayer(pid);
    engine.tutorialVote.removeVoter(pid);
    _broadcastLobby();
    if (engine.tutorialVote.isOpen || engine.tutorialVote.hasResult) {
      _broadcastTutorialState();
    }
    _emit();
  }

  void _handleJoin(GuestId guest, Map<String, Object?> json) {
    if (engine.phase != MafiaPhase.lobby) {
      _server.send(guest, jsonEncode({'type': 'error', 'message': 'Game already started'}));
      return;
    }
    final name = (json['name'] as String?)?.trim();
    if (name == null || name.isEmpty) return;
    final pid = 'g${_guestToPlayer.length + 1}';
    engine.addPlayer(id: pid, name: name);
    _guestToPlayer[guest] = pid;
    _playerToGuest[pid] = guest;
    _server.send(
        guest,
        jsonEncode({
          'type': 'welcome',
          'yourId': pid,
          'yourName': name,
        }));
    _broadcastLobby();
    _broadcastTutorialState();
    _emit();
  }

  void _applyNightAction(String playerId, String targetId) {
    final p = engine.players[playerId];
    if (p == null || !p.alive) return;
    bool ready = false;
    if (p.role == MafiaRole.mafia) {
      ready = engine.submitMafiaVote(playerId, targetId);
    } else if (p.role == MafiaRole.doctor) {
      ready = engine.submitDoctorTarget(playerId, targetId);
    }
    _emit();
    if (ready) {
      engine.resolveNight();
      _broadcastNightResult();
      _emit();
    }
  }

  void _applyDayVote(String playerId, String? targetId) {
    final ready = engine.submitDayVote(playerId, targetId);
    _broadcastVoteUpdate();
    _emit();
    if (ready) {
      engine.resolveDay();
      _broadcastDayResult();
      if (engine.phase == MafiaPhase.gameOver) {
        _broadcastGameOver();
      } else {
        _broadcastPhase();
      }
      _emit();
    }
  }

  // ---- Outbound -------------------------------------------------------

  void _broadcastLobby() {
    _server.broadcast(jsonEncode({
      'type': 'lobby',
      'players': engine.players.values
          .map((p) => {'id': p.id, 'name': p.name, 'isHost': p.isHost})
          .toList(),
      'canStart': engine.canStart,
    }));
  }

  // ---- Tutorial vote ---------------------------------------------------

  void _openTutorialVote() {
    if (engine.phase != MafiaPhase.lobby) return;
    if (engine.tutorialVote.isOpen) return;
    if (engine.tutorialVote.tutorialShown) return;
    engine.tutorialVote.open(engine.players.keys);
    _broadcastTutorialState();
    _emit();
  }

  void _submitTutorialVote(String voterId, bool yes) {
    if (!engine.tutorialVote.isOpen) return;
    engine.tutorialVote.submit(voterId, yes);
    _broadcastTutorialState();
    _emit();
  }

  void _broadcastTutorialState() {
    final v = engine.tutorialVote;
    final payload = <String, Object?>{
      'type': 'tutorial_vote_state',
      'isOpen': v.isOpen,
      'yesCount': v.yesCount,
      'noCount': v.noCount,
      'eligibleCount': v.eligibleCount,
      'result': v.result,
      'tutorialShown': v.tutorialShown,
    };
    if (v.result == true && !v.tutorialShown) {
      payload['title'] = GameTutorials.mafia.title;
      payload['sections'] = GameTutorials.mafia.sectionsJson();
      payload['menuSections'] = GameTutorials.mafia.browserMenuSectionsJson();
    }
    _server.broadcast(jsonEncode(payload));
  }

  void _sendRolesPrivately() {
    final mafiaIds = engine.players.values
        .where((p) => p.role == MafiaRole.mafia)
        .map((p) => p.id)
        .toList();
    for (final p in engine.players.values) {
      final payload = jsonEncode({
        'type': 'role',
        'role': p.role!.name,
        if (p.role == MafiaRole.mafia) 'mafiaIds': mafiaIds,
      });
      final guest = _playerToGuest[p.id];
      if (guest != null) _server.send(guest, payload);
    }
  }

  void _broadcastPhase() {
    _server.broadcast(jsonEncode({
      'type': 'phase',
      'phase': engine.phase.name,
      'day': engine.day,
      'alive': engine.alive.map(_publicPlayer).toList(),
      'dead': engine.dead.map(_publicPlayer).toList(),
    }));
  }

  void _broadcastVoteUpdate() {
    _server.broadcast(jsonEncode({
      'type': 'vote_update',
      'votes': engine.dayVotes.map((k, v) => MapEntry(k, v ?? '')),
    }));
  }

  void _broadcastNightResult() {
    final n = engine.lastNight!;
    _server.broadcast(jsonEncode({
      'type': 'phase',
      'phase': engine.phase.name,
      'day': engine.day,
      'alive': engine.alive.map(_publicPlayer).toList(),
      'dead': engine.dead.map(_publicPlayer).toList(),
      'killedId': n.killedId,
      'savedId': n.savedId,
    }));
  }

  void _broadcastDayResult() {
    final d = engine.lastDay!;
    _server.broadcast(jsonEncode({
      'type': 'day_result',
      'eliminatedId': d.eliminatedId,
      'tally': d.tally,
      'alive': engine.alive.map(_publicPlayer).toList(),
      'dead': engine.dead.map(_publicPlayer).toList(),
      'eliminatedRole': d.eliminatedId == null
          ? null
          : engine.players[d.eliminatedId]!.role!.name,
    }));
  }

  void _broadcastGameOver() {
    _server.broadcast(jsonEncode({
      'type': 'game_over',
      'winner': engine.winner!.name,
      'roles': {
        for (final p in engine.players.values) p.id: p.role!.name,
      },
    }));
  }

  Map<String, Object?> _publicPlayer(Player p) =>
      {'id': p.id, 'name': p.name, 'alive': p.alive};

  void _emit() => _stateChanges.add(null);
}

File: android/app/src/main/java/com/example/jamboree/games/mafia/MafiaServer.kt
package com.example.jamboree.games.mafia

import android.content.Context
import com.example.jamboree.join.LoopbackGuest
import com.example.jamboree.social.GuestId
import com.example.jamboree.social.HostServer
import com.example.jamboree.stats.StatsStore
import com.example.jamboree.tutorials.GameTutorials
import org.json.JSONArray
import org.json.JSONObject

/**
 * Wraps [HostServer] with Mafia-specific routing. Owns the engine, fans out
 * the right private/public messages, and converts incoming guest commands
 * into engine calls. Mirrors lib/games/mafia/mafia_server.dart.
 */
class MafiaServer(context: Context, val hostName: String = "Host") {
    var engine = MafiaEngine(); private set
    private val server = HostServer(html = HostServer.htmlAsset(context, "mafia_browser.html"), ctx = context)
    private val appCtx = context.applicationContext
    private val guestToPlayer = HashMap<GuestId, String>()
    private val playerToGuest = HashMap<String, GuestId>()

    var onStateChange: (() -> Unit)? = null
    private var statRecorded = false

    init {
        server.onMessage = ::onMessage
        server.onLeave = ::onLeave
    }

    companion object { const val HOST_ID = "host" }

    fun start(): String? {
        engine.addPlayer(HOST_ID, hostName, isHost = true)
        val url = server.startServer()
        // The host plays as a normal player on the same screen guests see.
        val local = server.attachLocalGuest()
        guestToPlayer[local] = HOST_ID
        playerToGuest[HOST_ID] = local
        emit()
        return url
    }

    /** In-process pipe for the host's own player screen. */
    fun makeLoopback(): LoopbackGuest = LoopbackGuest(server)

    fun stop() { server.stopServer(); resetState() }

    /** Clear all per-session state so the next time the host starts
     *  hosting they get a fresh screen — empty lobby, tutorial vote
     *  available again. */
    private fun resetState() {
        engine = MafiaEngine()
        guestToPlayer.clear(); playerToGuest.clear()
        statRecorded = false
        emit()
    }
    val guestCount: Int get() = server.guestCount

    // Host actions
    fun hostSetOptions(o: MafiaOptions) {
        engine.setOptions(o)
        broadcastOptions(); emit()
    }
    fun hostStart() {
        if (!engine.canStart) return
        engine.start()
        broadcastPhase(); sendRolesPrivately(); emit()
    }
    fun hostNightAction(targetId: String) = applyNightAction(HOST_ID, targetId)
    fun hostDayVote(targetId: String?) = applyDayVote(HOST_ID, targetId)
    fun hostCallTutorialVote() = openTutorialVote()
    fun hostTutorialVote(yes: Boolean) = submitTutorialVote(HOST_ID, yes)
    fun hostDismissTutorial() { engine.tutorialVote.markShown(); broadcastTutorialState(); emit() }
    fun advanceFromReveal() {
        engine.advanceToDayVote()
        if (engine.phase == MafiaPhase.GAME_OVER) broadcastGameOver() else broadcastPhase()
        emit()
    }

    // Inbound
    private fun onMessage(guest: GuestId, raw: String) {
        val j = runCatching { JSONObject(raw) }.getOrNull() ?: return
        when (j.optString("type")) {
            "join" -> handleJoin(guest, j)
            "night_action" -> guestToPlayer[guest]?.let { applyNightAction(it, j.getString("targetId")) }
            "vote" -> guestToPlayer[guest]?.let {
                applyDayVote(it, if (j.isNull("targetId")) null else j.getString("targetId"))
            }
            "chat" -> guestToPlayer[guest]?.let { relayChat(it, j.optString("text")) }
            // Only the host (not connected over WebSocket) may mutate options.
            "set_options" -> Unit
            "call_tutorial_vote" -> guestToPlayer[guest]?.let { openTutorialVote() }
            "tutorial_vote" -> guestToPlayer[guest]?.let { submitTutorialVote(it, j.getBoolean("yes")) }
        }
    }

    private fun onLeave(guest: GuestId) {
        val pid = guestToPlayer.remove(guest) ?: return
        playerToGuest.remove(pid)
        engine.removePlayer(pid)
        engine.tutorialVote.removeVoter(pid)
        broadcastLobby()
        if (engine.tutorialVote.isOpen || engine.tutorialVote.hasResult) broadcastTutorialState()
        emit()
    }

    private fun handleJoin(guest: GuestId, j: JSONObject) {
        if (engine.phase != MafiaPhase.LOBBY) {
            send(guest, JSONObject().put("type", "error").put("message", "Game already started"))
            return
        }
        val name = j.optString("name").trim().take(24)
        if (name.isEmpty()) return
        val pid = "p${guest.value}"
        engine.addPlayer(pid, name)
        guestToPlayer[guest] = pid
        playerToGuest[pid] = guest
        send(guest, JSONObject().put("type", "welcome").put("yourId", pid).put("yourName", name).put("game", "mafia"))
        broadcastLobby(); broadcastOptions(); broadcastTutorialState(); emit()
    }

    private fun broadcastOptions() {
        val o = engine.options
        broadcast(JSONObject()
            .put("type", "options")
            .put("mafiaCount", o.mafiaCount ?: JSONObject.NULL)
            .put("doctorEnabled", o.doctorEnabled)
            .put("maxMafiaCount", engine.maxMafiaCount))
    }

    private fun applyNightAction(playerId: String, targetId: String) {
        val p = engine.players[playerId] ?: return
        if (!p.alive) return
        val ready = when (p.role) {
            MafiaRole.MAFIA -> engine.submitMafiaVote(playerId, targetId)
            MafiaRole.DOCTOR -> engine.submitDoctorTarget(playerId, targetId)
            else -> false
        }
        emit()
        if (ready) { engine.resolveNight(); broadcastNightResult(); emit() }
    }

    private fun applyDayVote(playerId: String, targetId: String?) {
        val ready = engine.submitDayVote(playerId, targetId)
        broadcastVoteUpdate(); emit()
        if (ready) {
            engine.resolveDay()
            broadcastDayResult()
            if (engine.phase == MafiaPhase.GAME_OVER) broadcastGameOver() else broadcastPhase()
            emit()
        }
    }

    /** Relay a private message between mafia team-mates. Only a living mafia
     *  player may speak; every mafia (the whole "must agree" cabal) receives it. */
    private fun relayChat(playerId: String, text: String) {
        val sender = engine.players[playerId] ?: return
        if (sender.role != MafiaRole.MAFIA || !sender.alive) return
        val trimmed = text.trim().take(240)
        if (trimmed.isEmpty()) return
        val payload = JSONObject().put("type", "chat").put("fromId", sender.id)
            .put("fromName", sender.name).put("text", trimmed)
        for (p in engine.players.values) if (p.role == MafiaRole.MAFIA) {
            playerToGuest[p.id]?.let { send(it, payload) }
        }
    }

    // Outbound
    private fun broadcastLobby() {
        val arr = JSONArray()
        for (p in engine.players.values) {
            arr.put(JSONObject().put("id", p.id).put("name", p.name).put("isHost", p.isHost))
        }
        broadcast(JSONObject().put("type", "lobby").put("players", arr).put("canStart", engine.canStart))
    }

    private fun sendRolesPrivately() {
        val mafiaIds = JSONArray(engine.players.values.filter { it.role == MafiaRole.MAFIA }.map { it.id })
        for (p in engine.players.values) {
            val payload = JSONObject().put("type", "role").put("role", p.role!!.name.lowercase())
            if (p.role == MafiaRole.MAFIA) payload.put("mafiaIds", mafiaIds)
            playerToGuest[p.id]?.let { send(it, payload) }
        }
    }

    private fun broadcastPhase() {
        broadcast(JSONObject().apply {
            put("type", "phase"); put("phase", phaseJson(engine.phase)); put("day", engine.day)
            put("alive", publicArray(engine.alive)); put("dead", publicArray(engine.dead))
        })
    }
    private fun broadcastVoteUpdate() {
        val votes = JSONObject()
        for ((k, v) in engine.dayVotes) votes.put(k, v ?: "")
        broadcast(JSONObject().put("type", "vote_update").put("votes", votes))
    }
    private fun broadcastNightResult() {
        val n = engine.lastNight!!
        broadcast(JSONObject().apply {
            put("type", "phase"); put("phase", phaseJson(engine.phase)); put("day", engine.day)
            put("alive", publicArray(engine.alive)); put("dead", publicArray(engine.dead))
            put("killedId", n.killedId ?: JSONObject.NULL)
            put("savedId", n.savedId ?: JSONObject.NULL)
        })
    }
    private fun broadcastDayResult() {
        val d = engine.lastDay!!
        val role = d.eliminatedId?.let { engine.players[it]?.role?.name?.lowercase() }
        broadcast(JSONObject().apply {
            put("type", "day_result")
            put("eliminatedId", d.eliminatedId ?: JSONObject.NULL)
            put("tally", JSONObject(d.tally as Map<*, *>))
            put("alive", publicArray(engine.alive)); put("dead", publicArray(engine.dead))
            put("eliminatedRole", role ?: JSONObject.NULL)
        })
    }
    private fun broadcastGameOver() {
        if (!statRecorded) {
            statRecorded = true
            StatsStore.record(
                appCtx, "mafia",
                engine.players.values.map { it.name },
                if (engine.winner == MafiaWinner.TOWN) "town" else "mafia",
            )
        }
        val roles = JSONObject()
        for (p in engine.players.values) roles.put(p.id, p.role!!.name.lowercase())
        broadcast(JSONObject()
            .put("type", "game_over")
            .put("winner", if (engine.winner == MafiaWinner.TOWN) "town" else "mafia")
            .put("roles", roles))
    }

    private fun openTutorialVote() {
        if (engine.phase != MafiaPhase.LOBBY) return
        if (engine.tutorialVote.isOpen || engine.tutorialVote.tutorialShown) return
        engine.tutorialVote.open(engine.players.keys)
        broadcastTutorialState(); emit()
    }
    private fun submitTutorialVote(voterId: String, yes: Boolean) {
        if (!engine.tutorialVote.isOpen) return
        engine.tutorialVote.submit(voterId, yes)
        broadcastTutorialState(); emit()
    }
    private fun broadcastTutorialState() {
        val v = engine.tutorialVote
        val p = JSONObject()
            .put("type", "tutorial_vote_state")
            .put("isOpen", v.isOpen).put("yesCount", v.yesCount).put("noCount", v.noCount)
            .put("eligibleCount", v.eligibleCount)
            .put("result", v.result ?: JSONObject.NULL).put("tutorialShown", v.tutorialShown)
        if (v.result == true && !v.tutorialShown) {
            p.put("title", GameTutorials.mafia.title)
            p.put("sections", JSONArray(GameTutorials.mafia.sectionsJson().map { JSONObject(it as Map<*, *>) }))
            p.put("menuSections", JSONArray(GameTutorials.mafia.browserMenuSectionsJson().map { JSONObject(it as Map<*, *>) }))
        }
        broadcast(p)
    }

    private fun phaseJson(p: MafiaPhase): String = when (p) {
        MafiaPhase.LOBBY -> "lobby"; MafiaPhase.NIGHT -> "night"
        MafiaPhase.DAY_REVEAL -> "dayReveal"; MafiaPhase.DAY_VOTE -> "dayVote"
        MafiaPhase.GAME_OVER -> "gameOver"
    }

    private fun publicArray(ps: List<MafiaPlayer>): JSONArray {
        val arr = JSONArray()
        for (p in ps) arr.put(JSONObject().put("id", p.id).put("name", p.name).put("alive", p.alive))
        return arr
    }

    private fun emit() { onStateChange?.invoke() }
    private fun broadcast(obj: JSONObject) = server.broadcast(obj.toString())
    private fun send(guest: GuestId, obj: JSONObject) = server.send(guest, obj.toString())
}
